<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Issue 1136: Incomplete specification of nested_exception::rethrow_nested()</title>
<meta property="og:title" content="Issue 1136: Incomplete specification of nested_exception::rethrow_nested()">
<meta property="og:description" content="C++ library issue. Status: C++11">
<meta property="og:url" content="https://timsong-cpp.github.io/lwg-issues/1136.html">
<meta property="og:type" content="website">
<meta property="og:image" content="http://cplusplus.github.io/LWG/images/cpp_logo.png">
<meta property="og:image:alt" content="C++ logo">
<style>
  p {text-align:justify}
  li {text-align:justify}
  pre code.backtick::before { content: "`" }
  pre code.backtick::after { content: "`" }
  blockquote.note
  {
    background-color:#E0E0E0;
    padding-left: 15px;
    padding-right: 15px;
    padding-top: 1px;
    padding-bottom: 1px;
  }
  ins {background-color:#A0FFA0}
  del {background-color:#FFA0A0}
  table.issues-index { border: 1px solid; border-collapse: collapse; }
  table.issues-index th { text-align: center; padding: 4px; border: 1px solid; }
  table.issues-index td { padding: 4px; border: 1px solid; }
  table.issues-index td:nth-child(1) { text-align: right; }
  table.issues-index td:nth-child(2) { text-align: left; }
  table.issues-index td:nth-child(3) { text-align: left; }
  table.issues-index td:nth-child(4) { text-align: left; }
  table.issues-index td:nth-child(5) { text-align: center; }
  table.issues-index td:nth-child(6) { text-align: center; }
  table.issues-index td:nth-child(7) { text-align: left; }
  table.issues-index td:nth-child(5) span.no-pr { color: red; }
  @media (prefers-color-scheme: dark) {
     html {
        color: #ddd;
        background-color: black;
     }
     ins {
        background-color: #225522
     }
     del {
        background-color: #662222
     }
     a {
        color: #6af
     }
     a:visited {
        color: #6af
     }
     blockquote.note
     {
        background-color: rgba(255, 255, 255, .10)
     }
  }
</style>
</head>
<body>
<hr>
<h3 id="1136"><a href="lwg-defects.html#1136">1136</a>. Incomplete specification of <code>nested_exception::rethrow_nested()</code></h3>
<p><b>Section:</b> 17.9.8 <a href="https://timsong-cpp.github.io/cppwp/except.nested">[except.nested]</a> <b>Status:</b> <a href="lwg-active.html#C++11">C++11</a>
 <b>Submitter:</b> Daniel Kr&uuml;gler <b>Opened:</b> 2007-06-06 <b>Last modified:</b> 2016-01-28</p>
<p><b>Priority: </b>Not Prioritized
</p>
<p><b>View all other</b> <a href="lwg-index.html#except.nested">issues</a> in [except.nested].</p>
<p><b>View all issues with</b> <a href="lwg-status.html#C++11">C++11</a> status.</p>
<p><b>Discussion:</b></p>
<p>
It was recently mentioned in a newsgroup article
<a href="http://groups.google.de/group/comp.std.c++/msg/f82022aff68edf3d">http://groups.google.de/group/comp.std.c++/msg/f82022aff68edf3d</a>
that the specification of the member function <code>rethrow_nested()</code> of the
class <code>nested_exception</code> is incomplete, specifically it remains unclear
what happens, if member <code>nested_ptr()</code> returns a null value. In
17.9.8 <a href="https://timsong-cpp.github.io/cppwp/except.nested">[except.nested]</a> we find only the following paragraph related to that:
</p>
<blockquote><pre>
void rethrow_nested() const; // [[noreturn]]
</pre>
<blockquote><p>
-4- <i>Throws:</i> the stored exception captured by this <code>nested_exception</code> object.
</p></blockquote>
</blockquote>
<p>
This is a problem, because it is possible to create an object of
<code>nested_exception</code> with exactly such a state, e.g.
</p>
<blockquote><pre>
#include &lt;exception&gt;
#include &lt;iostream&gt;

int main() try {
  std::nested_exception e; // OK, calls current_exception() and stores it's null value
  e.rethrow_nested(); // ?
  std::cout &lt;&lt; "A" &lt;&lt; std::endl;
}
catch(...) {
  std::cout &lt;&lt; "B" &lt;&lt; std::endl;
}
</pre></blockquote>
<p>
I suggest to follow the proposal of the reporter, namely to invoke
<code>terminate()</code> if <code>nested_ptr()</code> return a null value of <code>exception_ptr</code> instead
of relying on the fallback position of undefined behavior. This would
be consistent to the behavior of a <code>throw;</code> statement when no
exception is being handled.
</p>

<p><i>[
2009 Santa Cruz:
]</i></p>


<blockquote><p>
Move to Ready.
</p></blockquote>



<p id="res-1136"><b>Proposed resolution:</b></p>
<p>
Change around 17.9.8 <a href="https://timsong-cpp.github.io/cppwp/except.nested">[except.nested]</a> p.4 as indicated:
</p>
<blockquote>
<p>
-4- <i>Throws:</i> the stored exception captured by this <code>nested_exception</code>
object<ins>, if <code>nested_ptr() != nullptr</code></ins>
</p>
<p>
<ins>- <i>Remarks:</i> If <code>nested_ptr() == nullptr</code>, <code>terminate()</code>
shall be called.</ins>
</p>
</blockquote>





</body>
</html>
